### PR TITLE
feat: support canisters and principals in dfx ledger account-id

### LINKED ISSUE #1981: Extend `dfx ledger account-id` to support canister accounts
Currently, developers need to write code to compute ledger account identifiers of their canisters.
This approach is error-prone and can result in lost ICP.
We need to extend the `dfx ledger account-id` command to support the computation of canister accounts.

Examples:
```sh
dfx ledger account-id --of-canister my_canister [--subaccount 000...]
<hex of a ledger account, canister alias is given>

dfx ledger account-id --of-principal 54yea-6qaaa-aaaaa-qaabq-cai [--subaccount 000...]
<hex of a ledger account, canister id is given>

dfx ledger account-id --of-principal sp3em-jkiyw-tospm-2huim-jor4p-et4s7-ay35f-q7tnm-hi4k2-pyicb-xae [--subaccount 0000...]
<hex of a user ledger account>
```

### PR DESCRIPTION
**Pull request by roman-kashitsyn**
This change adds new capabilities to the `dfx ledger account-id`
command:

  * The `--subaccount` option allows you to get an account id for a
    subaccount other than the default one.

  * The `--of-principal` option allows you to specify the controller
    of the subaccount to compute addresses that your identity doesn't
    control.

  * The `--of-canister` option allows you to specify canister alias
    instead of the principal.

Examples:

```sh
$ dfx ledger account-id --of-principal 53zcu-tiaaa-aaaaa-qaaba-cai
e9b0cdfcb858f5538d0d1d8d0a2700ccda299451c88bf686d00a3d8fb74dcfea

$ dfx ledger --network small02 account-id --of-canister ledger_demo
e9b0cdfcb858f5538d0d1d8d0a2700ccda299451c88bf686d00a3d8fb74dcfea

$ dfx ledger account-id \
  --of-principal 53zcu-tiaaa-aaaaa-qaaba-cai \
  --subaccount 0000000000000000000000000000000000000000000000000000000000000001
adddf401b1936409217493ed02b144ede539c6397c270b4ca8069f9930a4f4c9
```

This change also improves the error reporting by displaying error
causes if there are any. Example:

```sh
$ dfx ledger account-id \
  --of-principal 53zcu-tiaaa-aaaaa-qaaba-cai \
  --subaccount 00000000000000000000000000000000000000000000000000000000000000000
Error: Subaccount '00000000000000000000000000000000000000000000000000000000000000000' is not a valid hex string
Caused by:
  Odd number of digits
```

Fixes #1981.

### CODE REVIEW

**Review by ericswanson-dfinity — CHANGES_REQUESTED** [review]
Would you please add a note to the changelog?

**Comment by roman-kashitsyn** [comment]
> Would you please add a note to the changelog?

Good point, done!

File: src/dfx/src/main.rs [hunk 1]
```diff
@@ -125,7 +125,16 @@ fn main() {
         Err(e) => Err(e),
     };
     if let Err(err) = result {
-        eprintln!("{}", err);
+        for (level, cause) in err.chain().enumerate() {
+            if level == 0 {
+                eprintln!("Error: {}", err);
```
Comment (ericswanson-dfinity) [inline src/dfx/src/main.rs:130]: It looks like at least one of the e2e tests (failing in CI) will need to be updated to account for this change.

Comment (ericswanson-dfinity) [reply src/dfx/src/main.rs:130]: https://github.com/dfinity/sdk/runs/4869312230?check_suite_focus=true#step:8:101

**Review by ericswanson-dfinity — APPROVED** [review]

**Pull request merged**